[base] Codemod for hook directory migration

hbjORbj opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Follow-up on https://github.com/mui/material-ui/pull/36235#pullrequestreview-1307179392

mnajdova commented:
Should we wait for this to be merged before releasing? Having this in a release after people already migrated may not be as useful as having it in the same release. Can you please include a Breaking change description in https://github.com/mui/material-ui/pull/36235 so that people know how the migration should look like?

hbjORbj commented:
> Can you please include a Breaking change description in https://github.com/mui/material-ui/pull/36235 so that people know how the migration should look like?

@mnajdova Sure! I noticed you already wrote it in the release note. It looks good! Thanks for that.
```diff
-import { useBadge } from '@mui/base/BadgeUnstyled';
+import useBadge from '@mui/base/useBadge';
```

hbjORbj commented:
@michaldudak Thanks! I addressed your comment.

hbjORbj commented:
@michaldudak Thanks! I addressed your comment [here](https://github.com/mui/material-ui/pull/36295/commits/e5d7d5db587eddc974aae48120c254b6a47ce84f) and added the test [here](https://github.com/mui/material-ui/pull/36295/commits/e51721d8e6d60b3e0ff7c7aab9d803c8e940586b).

michaldudak reviewed (approved):
I tried some other scenarios and couldn't spot any more issues.

mnajdova commented on packages/mui-codemod/src/v5.0.0/base-hook-imports.js:
```diff
@@ -0,0 +1,113 @@
+      return sourceVal.startsWith('@mui/base/') && sourceVal.match('Unstyled');
+    }) // Process only Base components
+    .forEach((path) => {
+      // scenario 1: `@mui/base/<Component>Unstyled/use<Component>`
```
I don't think that this import was even supported at all. But let's keep it just in case 🤷

mnajdova commented on packages/mui-codemod/src/v5.0.0/base-hook-imports.js:
```diff
@@ -0,0 +1,113 @@
+  return str.charAt(0).toUpperCase() + str.slice(1);
+};
+
+const list = [
```
```suggestion
const movedExports = [
```

Let's give this list a more descriptive name.

mnajdova commented on packages/mui-codemod/src/v5.0.0/base-hook-imports.test/actual.js:
```diff
@@ -0,0 +1,8 @@
+import { useButton } from '@mui/base/ButtonUnstyled';
+import { useInput } from '@mui/base/InputUnstyled';
+import { useSelect, SelectOption, SelectChild, isOptionGroup, SelectOptionGroup } from '@mui/base/SelectUnstyled';
+import { UseOptionReturnValue } from '@mui/base/OptionUnstyled';
```
```suggestion
import { UseOptionReturnValue as OptionReturnValue } from '@mui/base/OptionUnstyled';
```

should we add test for aliasing the interfaces too?

mnajdova commented on packages/mui-codemod/src/v5.0.0/base-hook-imports.test/actual.js:
```diff
@@ -0,0 +1,8 @@
+import { useBadge } from '@mui/base/BadgeUnstyled';
+import { useButton } from '@mui/base/ButtonUnstyled';
+import { useInput } from '@mui/base/InputUnstyled';
+import { useSelect, SelectOption, SelectChild, isOptionGroup, SelectOptionGroup } from '@mui/base/SelectUnstyled';
```
```suggestion
import SelectUnstyled, { SelectUnstyledProps, useSelect, SelectOption, SelectChild, isOptionGroup, SelectOptionGroup } from '@mui/base/SelectUnstyled';
```

Let's ensure that the imports related to SelectUnstyled will not be lost.

hbjORbj commented:
@mnajdova Addressed your comments!
